runtime/src/main/java/org/corfudb/runtime/view/LayoutManagementView.java

```diff
@@ -434,43 +439,60 @@ private long getMaxGlobalTail(Layout layout, Layout.LayoutSegment segment) {
     public void reconfigureSequencerServers(Layout originalLayout, Layout newLayout,
                                             boolean forceReconfigure) {
 
-        long maxTokenRequested = 0L;
-        Map<UUID, Long> streamTails = Collections.emptyMap();
-        boolean bootstrapWithoutTailsUpdate = true;
-
-        // Reconfigure Primary Sequencer if required
-        if (forceReconfigure
-                || !originalLayout.getSequencers().get(0).equals(newLayout.getSequencers()
-                .get(0))) {
-            Layout.LayoutSegment latestSegment = newLayout.getSegments()
-                    .get(newLayout.getSegments().size() - 1);
-            maxTokenRequested = getMaxGlobalTail(newLayout, latestSegment);
-
-            FastObjectLoader fastObjectLoader = new FastObjectLoader(runtime);
-            fastObjectLoader.setRecoverSequencerMode(true);
-            fastObjectLoader.setLoadInCache(false);
-
-            // FastSMRLoader sets the logHead based on trim mark.
-            fastObjectLoader.setLogTail(maxTokenRequested);
-            fastObjectLoader.loadMaps();
-            streamTails = fastObjectLoader.getStreamTails();
-            verifyStreamTailsMap(streamTails);
-
-            // Incrementing the maxTokenRequested value for sequencer reset.
-            maxTokenRequested++;
-            bootstrapWithoutTailsUpdate = false;
-        }
-
-        // Configuring the new sequencer.
-        boolean sequencerBootstrapResult = CFUtils.getUninterruptibly(
-                runtime.getLayoutView().getRuntimeLayout(newLayout)
-                        .getPrimarySequencerClient()
-                        .bootstrap(maxTokenRequested, streamTails, newLayout.getEpoch(),
-                                bootstrapWithoutTailsUpdate));
-        if (sequencerBootstrapResult) {
-            log.info("Sequencer bootstrap successful.");
+        boolean acquiredLocked = recoverSequencerLock.tryLock();
+        if (acquiredLocked) {
+            try {
+
+                // Avoids stale bootstrap requests from being processed.
+                if (newLayout.getEpoch() <= lastKnownSequencerEpoch) {
+                    log.warn("Sequencer bootstrap failed. Already bootstrapped.");
+                    return;
+                }
+
+                long maxTokenRequested = 0L;
+                Map<UUID, Long> streamTails = Collections.emptyMap();
+                boolean bootstrapWithoutTailsUpdate = true;
+
+                // Reconfigure Primary Sequencer if required
+                if (forceReconfigure
+                        || !originalLayout.getSequencers().get(0).equals(newLayout.getSequencers()
+                        .get(0))) {
+                    Layout.LayoutSegment latestSegment = newLayout.getSegments()
+                            .get(newLayout.getSegments().size() - 1);
```

Comment:
You can actually fix the hole issue here, instead of only querying the tail query all logunits.

Reply:
I can but the new Runtimes will have the issue anyway. It would be cleaner to fix it during the stateTransfer.